File: python/gihyospark/ch05/basic-dataframe-example.py
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import sys
import warnings

import math
from datetime import date, datetime as dt

from pyspark import SparkConf, SparkContext
from pyspark.rdd import RDD
from pyspark.sql import Row, SQLContext
from pyspark.sql.functions import *
from pyspark.sql.readwriter import DataFrameWriter
from pyspark.sql.types import *
from pyspark.storagelevel import StorageLevel

class Station(object):
    """Station class"""

    def __init__(self, station_id, name, lat, lon, dockcount, landmark, installation):
        self.station_id   = station_id
        self.name         = name
        self.lat          = lat
        self.lon          = lon
        self.dockcount    = dockcount
        self.landmark     = landmark
        self.installation = installation

    @classmethod
    def is_head(cls, line):
        if "station_id" in line:
            return True
        else:
            return False

    @classmethod
    def parse(cls, line):
        elms = line.split(',')
        station_id = int(elms[0])
        name = elms[1]
        lat = float(elms[2])
        lon = float(elms[3])
        dockcount = int(elms[4])
        landmark = elms[5]
        parsed = dt.strptime(elms[6], "%m/%d/%Y")
        installation = date(parsed.year, parsed.month, parsed.day)
        return (station_id, name, lat, lon, dockcount, landmark, installation)


class Trip(object):

    def __init__(self, trip_id, duration,
                 startDate, startStation, startTerminal,
                 endDate, endStation, endTerminal,
                 bikeNum, subscriberType, zipcode):
        self.trip_id
        self.duration
        self.startDate
        self.startStation
        self.startTerminal
        self.endDate
        self.endStation
        self.endTerminal
        self.bikeNum
        self.subscriberType
        self.zipcode

    @classmethod
    def is_head(cls, line):
        if "Trip ID" in line:
            return True
        else:
            return False

    @classmethod
    def parse(cls, line):
        elms = line.split(',')
        trip_id = int(elms[0])
        duration = int(elms[1])
        startDate = dt.strptime(elms[2], "%m/%d/%Y %H:%M")
        startStation = elms[3]
        startTerminal = int(elms[4])
        endDate = dt.strptime(elms[5], "%m/%d/%Y %H:%M")
        endStation = elms[6]
        endTerminal = int(elms[7])
        bikeNum = int(elms[8])
        subscriberType = elms[9]
        zipcode = elms[10]
        return (trip_id, duration, startDate, startStation, startTerminal,
                endDate, endStation, endTerminal, bikeNum, subscriberType, zipcode)

class DataFrameExample(object):

    @classmethod
    def main(cls, params):
        if len(params) != 2:
            print("Usage: DataFrameExample <station_csv> <trip_csv>")
            exit(-1)
        station_path = params[0]
        trip_path = params[1]

        conf = SparkConf()
        sc = SparkContext(conf=conf)
        sqlContext = SQLContext(sc)
        example = DataFrameExample()
        example.run(sc, sqlContext, station_path, trip_path)

    def run(self, sc, sqlContext, station_path, trip_path):
        stationDF = self.read_station(sc, station_path)
        tripDF = self.read_trip(sc, trip_path)
        stationDF.cache()
        tripDF.cache()

        # Converts to RDD
        rdd = stationDF.rdd

        # Shows schema
        stationDF.printSchema()

        # Selects columns
        stationDF.select(stationDF.station_id).show(5, False)

        # Counts the number of records
        print(stationDF.count())

        # Filters records
        print(stationDF.filter(stationDF.landmark == "San Jose").count())

        # Sorts by a column
        stationDF.orderBy("dockcount", ascending=False).show(5, False)

        # Uses a function
        stationDF.select(stationDF.name, length("name"), expr("length(name)")).show(5, False)

        # Group By
        stationDF.groupBy(stationDF.dockcount).count().show(5, False)

        # Aggregation
        stationDF.groupBy(stationDF.landmark).agg({"dockcount": "sum"}).show()

        # Joins with DataFrames
        start = stationDF.alias("start") \
            .toDF("start_station_id", "start_name", "start_lat", "start_lon",
                 "start_dockcount", "start_landmark", "start_installation")
        end = stationDF.alias("end") \
            .toDF("end_station_id", "end_name", "end_lat", "end_lon",
                 "end_dockcount", "end_landmark", "end_installation")
        joinedTripDF = tripDF.join(start, tripDF.startTerminal == start.start_station_id, "inner") \
            .join(end, tripDF.endTerminal == end.end_station_id, "inner")
        joinedTripDF.printSchema()

        # Define an UDF
        def haversine_distance(pointA, pointB):
            deltaLat = math.radians(pointB[0] - pointA[0])
            deltaLon = math.radians(pointB[1] - pointA[1])
            a = math.pow(math.sin(deltaLat / 2), 2) + math.cos(math.radians(pointA[0])) * \
                math.cos(math.radians(pointB[0])) * math.pow(math.sin(deltaLon / 2), 2)
            greatCircleDistance = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
            return 6371000 * greatCircleDistance
        distance = udf(lambda latA, lonA, latB, lonB: haversine_distance((latA, lonA), (latB, lonB)),
                       DoubleType())
        joinedTripDF.select("startTerminal", 
                            "endTerminal", 
                            distance("start_lat", "start_lon", "end_lat", "end_lon")).show(5, False)

        # Writes a DataFrame
        joinedTripDF.write.mode("overwrite").format("json").save("./path/to/json/")

        # Reads a DataFrame
        loadedDF = sqlContext.read.format("json").load("./path/to/json/")
        loadedDF.printSchema()

        # Spark SQL
        tripDF.registerTempTable("trip")
        df = sqlContext \
            .sql("SELECT count(trip_id) AS count FROM trip WHERE to_date(startDate) = '2015-01-01'")
        df.show()

        # User an UDF in Spark SQL
        joinedTripDF.registerTempTable("joined_trip")
        sqlContext.registerFunction("distance",
                                    lambda latA, lonA, latB, lonB: \
                                    haversine_distance((latA, lonA), (latB, lonB)))
        result = sqlContext.sql("SELECT " +
            "startTerminal, " +
            "endTerminal, " +
            "distance(start_lat, start_lon, end_lat, end_lon) AS dist " +
            "FROM joined_trip " +
            "WHERE startTerminal != endTerminal")
        result.show(5, False)


    def read_station(self, sc, station_path):
        return sc.textFile(station_path, 2) \
            .filter(lambda line: not Station.is_head(line)) \
            .map(lambda line: Station.parse(line)) \
            .toDF().toDF("station_id", "name", "lat", "lon",
                         "dockcount", "landmark", "installation")

    def read_trip(self, sc, trip_path):
        return sc.textFile(trip_path, 2) \
            .filter(lambda line: not Trip.is_head(line)) \
            .map(lambda line: Trip.parse(line)) \
            .toDF().toDF( "trip_id", "duration",
                         "startDate", "startStation", "startTerminal",
                         "endDate", "endStation", "endTerminal",
                         "bikeNum", "subscriberType", "zipcode")


if __name__ == '__main__':
    DataFrameExample.main(sys.argv[1:])

File: src/main/scala/jp/gihyo/spark/ch05/BasicDataFrameExample.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.gihyo.spark.ch05

// scalastyle:off println
import java.text.SimpleDateFormat

import org.apache.spark.{SparkConf, SparkContext}
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.{SaveMode, SQLContext}
import org.apache.spark.sql.functions._

/**
  * An example code of the basic usage of DataFrame
  *
  * Run with
  * {{{
  * spark-submit --class jp.gihyo.spark.ch05.BasicDataFrameExample \
  *     path/to/gihyo-spark-book-example_2.10-1.0.1.jar
  *     201508_station_data.csv 201508_trip_data.csv
  * }}}
  */
object BasicDataFrameExample {

  def main(args: Array[String]): Unit = {
    if (args.length != 2) {
      new IllegalArgumentException("Invalid arguments")
      System.exit(1)
    }

    val conf = new SparkConf().setAppName("BasicDataFrameExample")
    val sc = new SparkContext(conf)
    val sqlContext = new SQLContext(sc)

    val stationPath = args(0)
    val tripPath = args(1)
    run(sc, sqlContext, stationPath, tripPath)

    sc.stop()
  }

  def run(
    sc: SparkContext,
    sqlContext: SQLContext,
    stationPath: String,
    tripPath: String): Unit = {
    import sqlContext.implicits._

    // Creates a DataFrame for the station data with a case class
    val stationRDD: RDD[(Int, String, Double, Double, Int, String, java.sql.Date)] =
      sc.textFile(stationPath).
        filter(line => !line.contains("station_id")).
        map { line =>
          val dateFormat = new SimpleDateFormat("MM/dd/yyy")

          val elms = line.split(",")
          val id = elms(0).toInt
          val name = elms(1)
          val lat = elms(2).toDouble
          val lon = elms(3).toDouble
          val dockcount = elms(4).toInt
          val landmark = elms(5)
          val parsedInstallation = dateFormat.parse(elms(6))
          val installation = new java.sql.Date(parsedInstallation.getTime)
          (id, name, lat, lon, dockcount, landmark, installation)
        }
    val stationDF = stationRDD.
      map { case (id, name, lat, lon, dockcount, landmark, installation) =>
        Station(id, name, lat, lon, dockcount, landmark, installation)
      }.toDF().cache()

    // Converts a DataFrame to RDD
    val rdd = stationDF.rdd
    // Uses getters of `Row`
    rdd.foreach { row =>
      val id = row.getInt(0)
      val name = row.getString(1)
      println(s"(id, name) = ($id, $name)")
    }
    rdd.foreach(row => println(s"${row.get(1)}"))

    // Shows the schema of the station DataFrame
    stationDF.printSchema()

    val localStation = stationDF.collect()

    // Selects and shows columns
    stationDF.select('id, $"name", stationDF("landmark"), col("dockcount")).show()
    stationDF.select('id, $"name", lit("STATION"), expr("id + 1")).show(5, false)

    // Counts the number of rows
    stationDF.count()
    stationDF.select('landmark).distinct().count()

    // Makes an alias of a column
    stationDF.select('name as "station_name").show(3, false)

    // Filters with a condition
    stationDF.filter('landmark === "San Jose").count()

    // Control flows
    stationDF.select('dockcount,
      when('dockcount >= 20, 2)
        .when('dockcount >= 15, 1)
        .otherwise(0)).show(3)


    // Sorts a DataFrame by columns or expressions
    stationDF.select('name, 'dockcount).orderBy('dockcount).show(5, false)
    stationDF.select('name, 'dockcount).orderBy('dockcount.desc).show(5, false)
    stationDF.select('name, 'lat, 'lon).orderBy('lat, 'lon).show(5, false)
    stationDF.select('name).orderBy(length('name)).show(5)

    // Uses a build-in function
    stationDF.select('name, length('name)).show(3, false)

    // Aggregation and grouping
    stationDF.agg(sum('dockcount)).show()
    stationDF.groupBy('landmark).count.show()
    stationDF.groupBy('landmark).agg(sum('dockcount)).show()
    stationDF.groupBy('landmark).agg(avg('lat), avg('lon)).show()

    // Create a DataFrame for the trip data
    val tripDF = sc.textFile(tripPath).
      filter(line => !line.contains("Trip ID")).
      map(_.split(",")).
      filter(_.size == 11).
      map { elms =>
        val dateFormat = new SimpleDateFormat("MM/dd/yyy HH:mm")

        val id = elms(0).toInt
        val duration = elms(1).toInt
        val startDate = new java.sql.Timestamp(dateFormat.parse(elms(2)).getTime)
        val startStation = elms(3)
        val startTerminal = elms(4).toInt
        val endDate = new java.sql.Timestamp(dateFormat.parse(elms(5)).getTime)
        val endStation = elms(6)
        val endTerminal = elms(7).toInt
        val bikeNum = elms(8).toInt
        val subscriberType = elms(9)
        val zipcode = elms(10)
        Trip(id, duration,
          startDate, startStation, startTerminal,
          endDate, endStation, endTerminal,
          bikeNum, subscriberType, zipcode)
      }.toDF().cache()

    tripDF.printSchema()

    // Joins the trip DataFrame with the station DataFrame
    var start = stationDF.as("start")
    var end = stationDF.as("end")
    start.columns.foreach { col => start = start.withColumnRenamed(col, s"start_${col}") }
    end.columns.foreach { col => end = end.withColumnRenamed(col, s"end_${col}") }
    val joinedTripDF = tripDF.
      join(start, tripDF("startTerminal") === start("start_id"), "inner").
      join(end, tripDF("endTerminal") === end("end_id"), "inner").cache()
    joinedTripDF.select('start_id, 'end_id).show(3)

    // Defines and uses an UDF
    val distance = udf((latA: Double, lonA: Double, latB: Double, lonB: Double) =>
      haversineDistance((latA, lonA), (latB, lonB)))
    joinedTripDF.select(
      'startStation,
      'endStation,
      distance('start_lat, 'start_lon, 'end_lat, 'end_lon) as "dist"
    ).filter('dist > 0.0).show(5, false)

    // Writes a DataFrame as JSON files
    tripDF.write.mode(SaveMode.Overwrite).format("json").save("./path/to/json/")

    // Loads a DataFrame from JSON files
    val df = sqlContext.read.format("json").load("./path/to/json/")

    // Operates a DataFrame with Spark SQL
    tripDF.registerTempTable("trip")
    val numRides = sqlContext.
      sql("SELECT count(id) AS count FROM trip WHERE to_date(startDate) = '2015-01-01'")
    numRides.show(1, false)

    // Registers an UDF on Spark SQL
    sqlContext.udf.register("distance",
      (lat1: Double, lon1: Double, lat2: Double, lon2: Double) =>
        haversineDistance((lat1, lon1), (lat2, lon2)))
    joinedTripDF.registerTempTable("joined_trip")
    val result = sqlContext.
      sql("SELECT " +
        "startTerminal, " +
        "endTerminal, " +
        "distance(start_lat, start_lon, end_lat, end_lon) AS dist " +
        "FROM joined_trip " +
        "WHERE startTerminal != endTerminal")
    result.show(3)

    // Unpersists DataFrames
    stationDF.unpersist()
    tripDF.unpersist()
    joinedTripDF.unpersist()
  }

  /**
    * Calculates a Haversine distance between two points by their latitudes and longitudes
    * SEE ALSO: https://en.wikipedia.org/wiki/Haversine_formula
    *
    * @param pointA Tuple2 of latitude and longitude
    * @param pointB Tuple2 of latitude and longitude
    * @return a Haversine distance
    */
  def haversineDistance(pointA: (Double, Double), pointB: (Double, Double)): Double = {
    val deltaLat = math.toRadians(pointB._1 - pointA._1)
    val deltaLon = math.toRadians(pointB._2 - pointA._2)
    val a = math.pow(
      math.sin(deltaLat / 2), 2) + math.cos(math.toRadians(pointA._1)) *
      math.cos(math.toRadians(pointB._1)) * math.pow(math.sin(deltaLon / 2), 2)
    val greatCircleDistance = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    6371000 * greatCircleDistance
  }
}

// scalastyle:on println
